adafruit_led_animation/animation.py

```diff
@@ -0,0 +1,585 @@
+                yield
+            if self.bounce:
+                self.reverse = not self.reverse
+
+    def draw(self):
+        next(self._generator)
+
+
+class Sparkle(Animation):
+    """
+    Sparkle animation of a single color.
+
+    :param pixel_object: The initialised LED object.
+    :param int speed: Animation speed in seconds, e.g. ``0.1``.
+    :param color: Animation color in ``(r, g, b)`` tuple, or ``0x000000`` hex format.
+    """
+    def __init__(self, pixel_object, speed, color):
+        self._half_color = None
+        self._dim_color = None
+        super(Sparkle, self).__init__(pixel_object, speed, color)
+
+    def _recompute_color(self, color):
+        half_color = tuple(color[rgb] // 4 for rgb in range(len(color)))
+        dim_color = tuple(color[rgb] // 10 for rgb in range(len(color)))
+        for pixel in range(len(self.pixel_object)):
+            if self.pixel_object[pixel] == self._half_color:
+                self.pixel_object[pixel] = half_color
+            elif self.pixel_object[pixel] == self._dim_color:
+                self.pixel_object[pixel] = dim_color
+        self._half_color = half_color
+        self._dim_color = dim_color
+
+    def draw(self):
+        pixel = random.randint(0, (len(self.pixel_object) - 2))
+        self.pixel_object[pixel] = self._color
+        self.show()
+        self.pixel_object[pixel] = self._half_color
+        self.pixel_object[pixel + 1] = self._dim_color
+        self.show()
+
+
+class Pulse(Animation):
+    """
+    Pulse all pixels a single color.
+
+    :param pixel_object: The initialised LED object.
+    :param int speed: Animation refresh rate in seconds, e.g. ``0.1``.
+    :param color: Animation color in ``(r, g, b)`` tuple, or ``0x000000`` hex format.
+    :param period: Period to pulse the LEDs over.  Default 5.
+    :param max_intensity: The maximum intensity to pulse, between 0 and 1.0.  Default 1.
+    :param min_intensity: The minimum intensity to pulse, between 0 and 1.0.  Default 0.
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, period=5, max_intensity=1, min_intensity=0):
+        self.max_intensity = max_intensity
+        self.min_intensity = min_intensity
+        self._period = period
+        self._intensity_delta = max_intensity - min_intensity
+        self._radians_per_second = radians(180 / period)
+        self._bpp = len(pixel_object[0])
+        self._last_update = monotonic_ns()
+        self._cycle_position = 0
+        super(Pulse, self).__init__(pixel_object, speed, color)
+
+    def draw(self):
+        now = monotonic_ns()
+        time_since_last_draw = (now - self._last_update) / 1000000000
+        self._last_update = now
+        self._cycle_position = (self._cycle_position + time_since_last_draw) % self._period
+        intensity = self.min_intensity + (
+            sin(self._radians_per_second * self._cycle_position) * self._intensity_delta)
```

Comment:
sin's range is -1.0 to 1.0, so this could end up setting intensity to a negative number.

For example:
```python
min_intensity = 0.0
max_intensity = 1.0
intensity_delta = max_intensity - min_intensity  # 1.0

# assume radians per second and cycle position end up passing 1.5pi to sin.
intensity = min_intensity + (sin(1.5 * pi) * intensity_delta)
intensity = 0.0 + (-1.0 * 1.0)
intensity = -1.0
```

The pixel colors would then be multiplied by `-1` which likely isn't what we want.

You can solve this by moving sin's range to be `0 - 1.0` like this:

```python
def normalized_sin(angle):
    return 0.5 + (sin(angle) * 0.5)
```

Reply:
I'm planning to remove the use of sin() very soon, as the performance is poor.  The code, however, shouldn't produce a negative brightness because it's clamped to 0-180 degrees by self._radians_per_second and % self._period set during init.

Reply:
Sin removed in favour of a linear function for now.  Code still needs significant optimization as  expected.